[BUG] Safety guard incorrectly blocks commands containing URLs

## Quick Summary

When executing commands containing URLs, the safety guard incorrectly matches the URL path portion as a file system path, causing the command to be blocked. For example, `agent-browser open https://github.com` is incorrectly blocked.

## Environment & Tools
- **PicoClaw Version:** b8f8e3f
- **Go Version:** go 1.26.1 linux/arm64
- **AI Model & Provider:** N/A
- **Operating System:** Linux 6.14.0-1018-oracle (aarch64)
- **Channels:** N/A

## 📸 Steps to Reproduce
1. Start PicoClaw with `restrictToWorkspace` safety restriction enabled
2. Execute a command containing a URL, e.g.: `agent-browser open https://github.com`
3. Observe the error in logs

## ❌ Actual Behavior

The command is blocked by the safety guard with the following error:

```
2026/03/07 03:31:32 [INFO] tool: Tool execution started {tool=exec, args=map[command:agent-browser open https://github.com]}
2026/03/07 03:31:32 [ERROR] tool: Tool execution failed {tool=exec, duration=0, error=Command blocked by safety guard (path outside working dir)}
```

## ✅ Expected Behavior

Commands containing URLs should execute normally without being blocked by the safety guard. URL path portions (e.g., `https://github.com/user/repo`) should not be treated as file system paths.

## 💬 Additional Context

**Root Cause:**

In `pkg/tools/shell.go`, the `absolutePathPattern` regex is used to match absolute file paths in commands:

```go
absolutePathPattern = regexp.MustCompile(`[A-Za-z]:\\[^\\\"']+|/[^\s\"']+`)
```

The Unix path portion `/[^\s\"']+` incorrectly matches URL path portions as absolute paths. For example:

- Input: `https://github.com`
- Match result: `//github.com` (mistakenly treated as a file path)

Since `/github.com` is not within the working directory, it triggers the "path outside working dir" error.

A fix will be submitted in a follow-up pull request.

--- a/pkg/tools/shell.go
+++ b/pkg/tools/shell.go
@@ -336,35 +336,34 @@ func (t *ExecTool) guardCommand(command, cwd string) string {
 			return ""
 		}
 
-		matches := absolutePathPattern.FindAllString(cmd, -1)
+		// Web URL schemes whose path components (starting with //) should be exempt
+		// from workspace sandbox checks. file: is intentionally excluded so that
+		// file:// URIs are still validated against the workspace boundary.
+		webSchemes := []string{"http:", "https:", "ftp:", "ftps:", "sftp:", "ssh:", "git:"}
 
-		for _, raw := range matches {
+		matchIndices := absolutePathPattern.FindAllStringIndex(cmd, -1)
+
+		for _, loc := range matchIndices {
+			raw := cmd[loc[0]:loc[1]]
+
 			// Skip URL path components that look like they're from web URLs.
 			// When a URL like "https://github.com" is parsed, the regex captures
 			// "//github.com" as a match (the path portion after "https:").
-			// These double-slash prefixes indicate URL paths, not file system paths.
-			// However, we must NOT skip file:// URIs as they could escape the sandbox.
-			// Only skip if preceded by a web URL scheme (http:, https:, ftp:, etc.).
-			if strings.HasPrefix(raw, "//") {
-				// Check if this // path is preceded by a web URL scheme
-				// by looking for patterns like "http://", "https://", "ftp://" before the match
-				idx := strings.Index(cmd, raw)
-				if idx > 0 {
-					// Look for the scheme prefix (e.g., "https:") before the //
-					before := cmd[:idx]
-					// Check if it ends with a web URL scheme followed by colon
-					// Web schemes: http, https, ftp, ftps, sftp, ssh, git
-					webSchemes := []string{"http:", "https:", "ftp:", "ftps:", "sftp:", "ssh:", "git:"}
-					isWebURL := false
-					for _, scheme := range webSchemes {
-						if strings.HasSuffix(before, scheme) {
-							isWebURL = true
-							break
-						}
+			// Use the exact match position (loc[0]) so that duplicate //path substrings
+			// in the same command are each evaluated at their own position.
+			if strings.HasPrefix(raw, "//") && loc[0] > 0 {
+				before := cmd[:loc[0]]
+				isWebURL := false
+
+				for _, scheme := range webSchemes {
+					if strings.HasSuffix(before, scheme) {
+						isWebURL = true
+						break
 					}
-					if isWebURL {
-						continue
-					}
+				}
+
+				if isWebURL {
+					continue
 				}
 			}
 

--- a/pkg/tools/shell_test.go
+++ b/pkg/tools/shell_test.go
@@ -515,3 +515,29 @@ func TestShellTool_FileURISandboxing(t *testing.T) {
 		}
 	}
 }
+
+// TestShellTool_URLBypassPrevented verifies that a command cannot bypass the workspace
+// sandbox by smuggling a real path after a URL that contains the same //path substring.
+// e.g. "echo https://etc/passwd && cat //etc/passwd" must still be blocked.
+func TestShellTool_URLBypassPrevented(t *testing.T) {
+	tmpDir := t.TempDir()
+	tool, err := NewExecTool(tmpDir, true)
+	if err != nil {
+		t.Fatalf("unable to configure exec tool: %s", err)
+	}
+
+	// The path //etc/passwd appears twice: once as the host part of an https URL
+	// and once as a real (escaped) absolute path. The guard must block the command
+	// because the second occurrence is a genuine out-of-workspace path.
+	blockedCommands := []string{
+		"echo https://etc/passwd && cat //etc/passwd",
+		"curl https://host/file && ls //etc",
+	}
+
+	for _, cmd := range blockedCommands {
+		result := tool.Execute(context.Background(), map[string]any{"command": cmd})
+		if !result.IsError || !strings.Contains(result.ForLLM, "path outside working dir") {
+			t.Errorf("bypass attempt should be blocked: %q\n  got: %s", cmd, result.ForLLM)
+		}
+	}
+}

